### PR TITLE
Slight performance improvment for MutableHandlerRegistryImpl

### PR DESCRIPTION
**Pull request by nmittler**
This is to help address #318

@ejona86 @buchgr PTAL

### CODE REVIEW

**Comment by nmittler** [comment]
Here are the benchmark results from running on my laptop:

```
Benchmark                              (methodCountPerService)  (nameLength)  (serviceCount)  (useNewRegistry)   Mode  Cnt    Score    Error  Units
HandlerRegistryBenchmark.lookupMethod                      100            50             100             false  thrpt   20  332.830 ±  9.159  ops/s
HandlerRegistryBenchmark.lookupMethod                      100            50             100              true  thrpt   20  377.629 ± 12.240  ops/s
```

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        Random r = new Random();
+        byte[] bytes = new byte[nameLength];
+        for (int ix = 0; ix<nameLength; ++ix) {
+            int charIx = r.nextInt(VALID_CHARACTERS.length());
+            bytes[ix] = (byte) VALID_CHARACTERS.charAt(charIx);
+        }
+        return new String(bytes);
+    }
+
+    private final class OldRegistry extends MutableHandlerRegistry {
```
Comment (ejona86) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: Can we not add the old registry, and instead just run the benchmark before and after the change?

Comment (nmittler) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: Sure ... I guess I've already done that so I can safely remove.

Comment (nmittler) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: done.

**Comment by nmittler** [comment]
@ejona86 PTAL ... I've removed the old registry.

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
+package io.grpc.benchmarks.netty;
```
Comment (ejona86) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: copyright header

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    public void lookupMethod(Blackhole bh) {
+        for (String fullMethodName : fullMethodNames) {
+            bh.consume(registry.lookupMethod(fullMethodName));
+        }
+    }
+
+    private String randomString() {
+        Random r = new Random();
+        byte[] bytes = new byte[nameLength];
+        for (int ix = 0; ix<nameLength; ++ix) {
```
Comment (ejona86) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: I'm almost certain than checkstyle will complain about this.

File: lib/netty [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit 871ce43b1f431536e08a36d420875494ef3e9451
+Subproject commit 74627483d7bdcae3130b2ba173e08ec5f2658377
```
Comment (ejona86) [inline lib/netty]: Revert?

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    @Benchmark
+    public void lookupMethod(Blackhole bh) {
+        for (String fullMethodName : fullMethodNames) {
+            bh.consume(registry.lookupMethod(fullMethodName));
+        }
+    }
+
+    private String randomString() {
+        Random r = new Random();
+        byte[] bytes = new byte[nameLength];
```
Comment (ejona86) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: Why not a char[]?

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java (hunk 3 above)
Comment (nmittler) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: done.

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java (hunk 5 above)
Comment (nmittler) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: done.

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java (hunk 2 above)
Comment (nmittler) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: done.

**Comment by nmittler** [comment]
@ejona86 PTAL

**Comment by buchgr** [comment]
@nmittler nice! I am without a Laptop right now, so can't  really look at the code, but the main concern was that it allocates quite a few string objects. So it's even more of a win that it also executes faster :-) +1

**Comment by ejona86** [comment]
@buchgr, it gets rid of one of the allocations; we still have one left.

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.benchmarks.netty;
+
+import io.grpc.MutableHandlerRegistryImpl;
+import io.grpc.ServerMethodDefinition;
+import io.grpc.ServerServiceDefinition;
+import org.openjdk.jmh.annotations.Benchmark;
```
Comment (ejona86) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: Insert newline above this line.

**Comment by ejona86** [comment]
@nmittler LGTM

File: benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java (hunk 6 above)
Comment (nmittler) [inline benchmarks/src/jmh/java/io/grpc/benchmarks/netty/HandlerRegistryBenchmark.java]: done.

**Comment by nmittler** [comment]
cherry-picked as 518b7db

**Pull request closed without merging**